cpu/stm32/periph/eth.c

```diff
@@ -650,20 +651,28 @@ static int stm32_eth_recv(netdev_t *netdev, void *buf, size_t max_len,
         return -ENOBUFS;
     }
 
+    /* Fetch payload, collect RX timestamp from last descriptor if module periph_ptp is used, and
+     * hand DMA descriptors back to the DMA */
     size_t remain = size;
-    while (remain) {
-        size_t chunk = MIN(remain, ETH_RX_BUFFER_SIZE);
-        memcpy(data, rx_curr->buffer_addr, chunk);
-        data += chunk;
-        remain -= chunk;
-        /* Hand over descriptor to DMA */
-        rx_curr->status = RX_DESC_STAT_OWN;
-        rx_curr = rx_curr->desc_next;
-    }
-
-    if ((size + ETHERNET_FCS_LEN - 1) % ETH_RX_BUFFER_SIZE < ETHERNET_FCS_LEN) {
-        /* one additional rx descriptor was needed only for the FCS, hand that
-         * back to the DMA as well */
+    while (1) {
+        /* there can be more DMA descriptors than needed for holding the Ethernet
+         * payload, as the 4 byte FCS will also be stored by DMA */
+        if (remain) {
+            size_t chunk = MIN(remain, ETH_RX_BUFFER_SIZE);
+            memcpy(buf, rx_curr->buffer_addr, chunk);
+            buf += chunk;
+            remain -= chunk;
+        }
+        if (rx_curr->status & RX_DESC_STAT_LS) {
```

Comment:
What is `RX_DESC_STAT_LS`?

```suggestion
        /* last element */
        if (rx_curr->status & RX_DESC_STAT_LS) {
```

Reply:
It is set by the DMA on the last chunk of one Ethernet Frame. Previously, the driver relied on the size for terminating the loop (plus some special case handling for when the 4 byte FCS overflowed the last chunk, resulting in one DMA buffer more being used). This should actually be more bullet proof, at least when trusting the DMA to correctly mark the last chunk.

Reply:
Just add a short note, makes this block easier to understand on the first read.